clippy_lints/src/undocumented_unsafe_block_safety.rs

```diff
@@ -0,0 +1,198 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::{snippet, indent_of, reindent_multiline};
+use rustc_errors::Applicability;
+use rustc_hir::{Block, BlockCheckMode, ExprKind, HirId, Local, UnsafeSource};
+use rustc_lexer::TokenKind;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty::TyCtxt;
+use rustc_session::{impl_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+use core::str::Lines;
+use core::iter::Rev;
+use std::borrow::Cow;
+use clippy_utils::is_lint_allowed;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `unsafe` blocks without a `// Safety: ` comment
+    /// explaining why the unsafe operations performed inside
+    /// the block are (or are not) safe.
+    ///
+    /// ### Why is this bad?
+    /// Unsafe blocks without a safety comment can be difficult to
+    /// understand for those that are unfamiliar with the block's code
+    /// and/or unsafe in general.
+    ///
+    /// ### Example
+    /// ```rust
+    ///   unsafe {};
+    ///
+    ///   let _ = unsafe {};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // Safety: ...
+    /// unsafe {}
+    ///
+    /// // Safety: ...
+    /// let _ = unsafe {};
+    /// ```
+    pub UNDOCUMENTED_UNSAFE_BLOCK_SAFETY,
+    restriction,
+    "creating an unsafe block without explaining why it is or is not safe"
+}
+
+impl_lint_pass!(UndocumentedUnsafeBlockSafety => [UNDOCUMENTED_UNSAFE_BLOCK_SAFETY]);
+
+#[derive(Default)]
+pub struct UndocumentedUnsafeBlockSafety {
+    pub local: bool
+}
+
+#[derive(PartialEq)]
+enum CommentType {
+    Safety,
+    Unrelated,
+    None,
+}
+
+impl LateLintPass<'_> for UndocumentedUnsafeBlockSafety {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &'_ Block<'_>) {
+        if self.local {
+            self.local = false;
```

Comment:
That will produce bugs:

```rs
let _ = {
    let _ = {
        // ...
    }
    // self.local is false, even though we're still in a local.
}
```

Reply:
I just tested this out:
```rust
    let _ = {
        let _ = {
            let _ = unsafe {};
        };
    };
```
```
error: unsafe block missing a safety comment
  --> $DIR/undocumented_unsafe_blocks.rs:8:13
   |
LL |             let _ = unsafe {};
   |             ^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::undocumented-unsafe-blocks` implied by `-D warnings`
```
It still only ends up checking the local with the unsafe block.

Reply:
The problem is not the inner unsafe block, but if you put an unsafe block after leaving one of the inner blocks. We have this situation multiple times. For this we usually count how many layers deep we are with `saturating_add(1)` and `saturating_sub(1)`, instead of setting a boolean.